### PR TITLE
Accessing magnitudes from COSMOSCatalog

### PR DESCRIPTION
**Pull request by arunkannawadi**
After some of the recent changes made to COSMOSCatalog, the way to access the magnitudes of the entries appears to have changed. This is a 2-line PR addressing those changes in demo13.py, which would not run otherwise.

### CODE REVIEW

File: examples/demo13.py [hunk 1]
```diff
@@ -174,7 +174,7 @@ def main(argv):
         # Note that we could use wcs.toWorld() to get the (RA, dec) for these (x, y) positions.  Or,
         # if we had started with (RA, dec) positions, we could have used wcs.toImage() to get the
         # CCD coordinates for those positions.
-        mag_stamp.append(cat.param_cat.mag_auto[pos_rng()*cat.nobjects])
+        mag_stamp.append(cat.param_cat[pos_rng()*cat.nobjects][1])
```
Comment (rmjarvis) [inline examples/demo13.py]: Don't use `[1]`.  You should always avoid magic numbers in code.  The better fix here is `param_cat['mag_auto'][...]`.

Also, I'd add an `int(...)` for the random number in the argument, since currently it's a float, which numpy must be converting to an `int` automatically, but it would be more intuitive to do that directly here, rather than rely on this kind of non-obvious feature.

**Comment by rmjarvis** [comment]
One other fix that needs to be made is line 196 still uses `deep=True`, which is inappropriate with the new default catalog.  I'm getting the warning:
```
/Library/Python/2.7/site-packages/galsim/scene.py:544: UserWarning: Ignoring `deep` argument, because the sample being used already corresponds to a flux limit of F814W<25.2
  'corresponds to a flux limit of F814W<25.2')
```
when I run demo13 now.  So probably ought to fix this here as well.

And thanks for catching this error.  I don't always run demo12 and 13 when I'm testing, so I apologize for missing this when we worked on COSMOSCatalog.

File: examples/demo13.py [hunk 2]
```diff
@@ -106,9 +106,9 @@ def main(argv):
     # something larger (but it should be <=100, the number of galaxies in this small example
     # catalog).  With 4000 galaxies in a 4k x 4k image with the WFIRST pixel scale, the effective
     # galaxy number density is 74/arcmin^2.  This is not the number density that is expected for a
-    # sample that is so bright (I<23.5) but it makes the image more visually interesting.  One could
+    # sample that is so bright (I<25.2) but it makes the image more visually interesting.  One could
```
Comment (arunkannawadi) [inline examples/demo13.py]: This was 23.5 initially, and I changed 23.5 to 25.2 everywhere. The number density will have to change and probably the comment itself.

**Comment by arunkannawadi** [comment]
No problem @rmjarvis. I had to run this script to cross check with
something that I am doing, which is when I recognized this error. I have
made the changes you mentioned.

On Sun, May 1, 2016 at 4:37 PM, Mike Jarvis <notifications@github.com>
wrote:

> One other fix that needs to be made is line 196 still uses deep=True,
> which is inappropriate with the new default catalog. I'm getting the
> warning:
>
> /Library/Python/2.7/site-packages/galsim/scene.py:544: UserWarning: Ignoring `deep` argument, because the sample being used already corresponds to a flux limit of F814W<25.2
>   'corresponds to a flux limit of F814W<25.2')
>
> when I run demo13 now. So probably ought to fix this here as well.
>
> And thanks for catching this error. I don't always run demo12 and 13 when
> I'm testing, so I apologize for missing this when we worked on
> COSMOSCatalog.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/GalSim-developers/GalSim/pull/744#issuecomment-216070395>
>



-- 
Regards

Arun Kannawadi Jayaraman
PhD candidate, Dept. of Physics,
Carnegie Mellon University,
Pittsburgh, PA 15213

File: examples/demo13.py [hunk 3]
```diff
@@ -93,11 +93,11 @@ def main(argv):
     logger.debug('Read in WFIRST imaging filters.')
 
     logger.info('Reading from a parametric COSMOS catalog.')
-    # Read in a galaxy catalog - just a random subsample of 100 galaxies for F814W<23.5 from COSMOS.
+    # Read in a galaxy catalog - just a random subsample of 100 galaxies for F814W<25.2 from COSMOS.
```
Comment (rmandelb) [inline examples/demo13.py]: Why is this changed?  Did you locally change the contents of ```real_galaxy_catalog_example_fits.fits``` to correspond to <25.2?  I don't see the change here.  If you didn't change the file, then all changes to the comments here are incorrect.

Comment (rmjarvis) [inline examples/demo13.py]: I assumed from the warning that was being emitted for `deep=True` that you had updated the example catalog.  If not, then there is an error in how the code is setting `use_sample` for this file name.

Comment (rmandelb) [inline examples/demo13.py]: I don't see ```use_sample``` being used for this at all?  Am I looking in the wrong place?

Comment (rmjarvis) [inline examples/demo13.py]: I think if you put back the deep=True line, you'll see what I mean.  `use_sample` gets set in COSMOSCatalog by the line:
```
full_file_name, full_image_dir, _, self.use_sample = \
            galsim.real._parse_files_dirs(file_name, image_dir, dir, noise_dir, sample)
```
This comes back with `use_sample` = None, which triggers a warning when `deep` is True.  If you didn't actually update the example file, then the logic there needs to change.

Comment (rmandelb) [inline examples/demo13.py]: Oh I thought you meant in this demo (that the catalog was being initialized using a ```sample``` name).  I see what you mean now.

Comment (arunkannawadi) [inline examples/demo13.py]: It essentially boils down to whether the example catalog file containing
100 galaxies that this demo loads was changed when the COSMOSCatalog was
updated. I have changed the comments in few other places too to say that we
get galaxies brighter than 25.2.
I think the maximum magnitude in the example catalog comes out to be under
22.5, so it really doesn't make the difference.

On Wed, May 4, 2016 at 1:21 PM, Rachel Mandelbaum <notifications@github.com>
wrote:

> In examples/demo13.py
> <https://github.com/GalSim-developers/GalSim/pull/744#discussion_r62078875>
> :
>
> > @@ -93,11 +93,11 @@ def main(argv):
> >      logger.debug('Read in WFIRST imaging filters.')
> >
> >      logger.info('Reading from a parametric COSMOS catalog.')
> > -    # Read in a galaxy catalog - just a random subsample of 100 galaxies for F814W<23.5 from COSMOS.
> > +    # Read in a galaxy catalog - just a random subsample of 100 galaxies for F814W<25.2 from COSMOS.
>
> Oh I thought you meant in this demo (that the catalog was being
> initialized using a sample name). I see what you mean now.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/GalSim-developers/GalSim/pull/744/files/356cf2f039aa391c0c6c6a2a1f9093d50ef29507#r62078875>
>



-- 
Regards

Arun Kannawadi Jayaraman
PhD candidate, Dept. of Physics,
Carnegie Mellon University,
Pittsburgh, PA 15213

Comment (rmandelb) [inline examples/demo13.py]: No, I did not update the example catalog (no reason to - COSMOSCatalog has multiple possible samples it can take, and there's no reason why the example has to be updated just because the default COSMOSCatalog choice changed).  I thought the changes in these comments meant you had updated it yourself and then not put the update in the repo.  As I said, if you didn't do that, then your changes to the comments should be reverted.

**Comment by arunkannawadi** [comment]
I think this is ready to be merged.

**Comment by rmandelb** [comment]
Sure, I think it's fine.

**Comment by rmjarvis** [comment]
Isn't the bug with using `deep=True` for the example great3 file still unfixed?  I think it should be allowed to use the deep option with that catalog, but that is now broken.  (It was prior to this PR -- this PR is just where we noticed it.)

**Comment by arunkannawadi** [comment]
I have removed the `deep=True` part from line 196. The demo file must run without any warnings or error.

**Comment by rmjarvis** [comment]
> The demo file must run without any warnings or error.

Sure.  But that's just avoiding the bug, not fixing it.  Since the example file is really I < 23.5, it should be valid to request `deep=True` with it.  But it raises an exception if you try.  That bug should be fixed.

**Comment by rmandelb** [comment]
Fair point.  Right now what the code does is:

- Check if you're using the <23.5 sample that is available on the GREAT3 site: if yes, then allow use of ```deep=True```.
- Check if you're using the <25.2 sample that is available on the GREAT3 site: if yes, then forbid use of ```deep=True```.

It doesn't allow the possibility that for user-defined samples, one might want to use ```deep=True```.  We should probably allow that, while documenting (in the docstring) that it's only meant for a particular case and won't do what the user intends for a completely arbitrary case.

**Comment by rmjarvis** [comment]
One option would be to change the example file name to have 23.5 in it to make that explicit.  Then have the code check for 23.5 in the filename at the place where it currently sets `use_sample=None` in real.py.

i.e. rather than bail out if `sample is None` and `file_name is not None`, look to see if there is a clue in the file name:
```
    if sample is None:
        if file_name is None:
            use_sample = '25.2'
        elif '23.5' in file_name:
            use_sample = '23.5'
        elif '25.2' in file_name:
            use_sample = '25.2'
        else:
            use_sample = None
    else:
        .... 
```
That's probably the sanest option in the presence of arbitrary user files.

**Comment by rmjarvis** [comment]
Also, the error message in scene.py should be updated for the case of `use_sample == None`, since right now the message implies that the catalog is already 25.2, whereas in the None case, it's really that we don't know what the limit is to properly simulate the 25.2 deep sample.

**Comment by rmandelb** [comment]
@arunkannawadi , what is the status of these final items?  Will this be ready for final review and merging soon?

**Comment by arunkannawadi** [comment]
I have incorporated Mike's suggestion about including 23.5 in the file name and looking for the flux limit in the file name itself.

**Comment by arunkannawadi** [comment]
Should the change in the file names go on CHANGELOG?

**Comment by rmjarvis** [comment]
> Should the change in the file names go on CHANGELOG?

Yes.

**Comment by arunkannawadi** [comment]
Done.

**Comment by rmjarvis** [comment]
I think this is ok to merge.

**Comment by arunkannawadi** [comment]
Great. I'll wait to hear from @rmandelb and merge it once she is okay with
it as well or tomorrow night, whichever is the latter.

On Mon, May 23, 2016 at 7:30 PM, Mike Jarvis <notifications@github.com>
wrote:

> I think this is ok to merge.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/GalSim-developers/GalSim/pull/744#issuecomment-221125775>
>



-- 
Regards

Arun Kannawadi Jayaraman
PhD candidate, Dept. of Physics,
Carnegie Mellon University,
Pittsburgh, PA 15213

**Comment by rmandelb** [comment]
I'm getting a bunch of unit test failures having to do with tests that use the example catalogs.

**Comment by arunkannawadi** [comment]
Sorry, I missed changing the example file names in the unit tests as well.
They now pass on my system.

On Mon, May 23, 2016 at 10:52 PM, Rachel Mandelbaum <
notifications@github.com> wrote:

> I'm getting a bunch of unit test failures having to do with tests that use
> the example catalogs.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/GalSim-developers/GalSim/pull/744#issuecomment-221153196>
>



-- 
Regards

Arun Kannawadi Jayaraman
PhD candidate, Dept. of Physics,
Carnegie Mellon University,
Pittsburgh, PA 15213

**Comment by rmandelb** [comment]
I guess you may not have committed and/or pushed all the files, because I just pulled and got some changes, but there are still tests that fail in ```test_config_gsobject.py``` and ```test_deprecated.py``` when I do ```scons tests```.  (Or if you didn't change them, then I don't understand how the tests could have passed on your system.)

**Comment by arunkannawadi** [comment]
I didn't execute `run_all_tests` on my laptop due to memory restrictions -
I had only explicitly checked those that I thought would have changed. I
did a more thorough search and renamed the files in many other files.
Please let me know if the tests pass.

On Tue, May 24, 2016 at 9:14 AM, Rachel Mandelbaum <notifications@github.com
> wrote:

> I guess you may not have committed and/or pushed all the files, because I
> just pulled and got some changes, but there are still tests that fail in
> test_config_gsobject.py and test_deprecated.py when I do scons tests. (Or
> if you didn't change them, then I don't understand how the tests could have
> passed on your system.)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/GalSim-developers/GalSim/pull/744#issuecomment-221264275>
>



-- 
Regards

Arun Kannawadi Jayaraman
PhD candidate, Dept. of Physics,
Carnegie Mellon University,
Pittsburgh, PA 15213

**Comment by rmjarvis** [comment]
`scons tests` is the normal way to run all the tests, and there shouldn't be significant memory issues running it that way.  `run_all_tests` actually means run *all* the tests including some that take more time and occasionally more memory.

You should always check that `scons tests` succeeds before pushing any new commits.

**Comment by arunkannawadi** [comment]
Thanks. I'll keep that in mind.

On Tue, May 24, 2016 at 12:59 PM, Mike Jarvis <notifications@github.com>
wrote:

> scons tests is the normal way to run all the tests, and there shouldn't
> be significant memory issues running it that way. run_all_tests actually
> means run *all* the tests including some that take more time and
> occasionally more memory.
>
> You should always check that scons tests succeeds before pushing any new
> commits.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/GalSim-developers/GalSim/pull/744#issuecomment-221336132>
>



-- 
Regards

Arun Kannawadi Jayaraman
PhD candidate, Dept. of Physics,
Carnegie Mellon University,
Pittsburgh, PA 15213

**Comment by arunkannawadi** [comment]
The Python tests don't seem to be running at all when I type in `scons tests`. Here's a part from the output:
```
Starting python tests...

----------------------------------------------------------------------
Ran 0 tests in 0.016s

OK
Nosetests finished successfully.

Starting cpp tests...
Running 12 test cases...
```

**Comment by rmjarvis** [comment]
@arunkannawadi  That's pretty weird.  If you type `nosetests` in the tests directory, does something sensible happen?  Can you normally run `scons tests` on the master branch, say?  If not, my guess is you have something set up wrong, but I don't recall seeing this particular symptom before, so I don't have a good guess as to what it might be.

On the other hand, I don't think we need you to figure this out to merge this particular branch.  It all seems to be working as far as I can tell, so I'm ok with merging this now.
